coredns: support IPv6 record set

pmichali opened the pull request:
Added support for AAAA record for coredns and included unit test.
Refactored common test code to reduce duplication from added test and
existing tests.
Fixed function names in comments for Google and AWS tests to match
actual test name in this area.

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#pull-request-process and developer guide https://github.com/kubernetes/community/blob/master/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/community/blob/master/contributors/devel/faster_reviews.md
3. Follow the instructions for writing a release note: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#release-notes
-->

**What this PR does / why we need it**:

Adding IPv6 support to kubernetes, once piece at a time. :)

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #44351

**Special notes for your reviewer**:
In addition to the change and unit test method, I did some minor refactoring (since the UT was a near clone of an existing test). Fixed typos in related test methods' comment lines. Please let me know if this is OK (I was thinking it was a small change, but don't know the protocol here), or if I need to break it into multiple commits.

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```NONE
```

pmichali committed:
coredns: support IPv6 record set

Added support for AAAA record for coredns and included unit test.

Fixed function names in comments for Google and AWS tests to match
actual test name in this area.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44403)
<!-- Reviewable:end -->

grodrigues3 commented:
@k8s-bot ok to test

quinton-hoole commented on federation/pkg/dnsprovider/providers/coredns/rrsets.go:
```diff
@@ -75,6 +75,8 @@ func (rrsets ResourceRecordSets) Get(name string) (dnsprovider.ResourceRecordSet
 			rrset.rrsType = rrstype.CNAME
 		case ip.To4() != nil:
 			rrset.rrsType = rrstype.A
+		default:
```
Do we not want to test ip.To6 here, instead of defaulting to ipv6?

pmichali replied:
I was thinking it wasn't necessary, because it looks like ParseIP will either eval the string to an IP (v4 or v6) or return nil. As a result, I figured that there are cases for nil and IPv4, so the value would have to be Ipv6, if those two cases didn't match. Is that acceptable?

pmichali replied:
Should I:

- Leave it as-is
- Add a comment indicating the default is for IPv6
- Add a case with test for To6(), and then have default case with comment indicating it should never occur?

pmichali replied:
Will take a guess at option #2, adding a comment, so that this can move forward.

pmichali commented:
Any feedback on the latest change set?

pmichali commented:
Change is ready for review - please.

shashidharatd commented:
Reviewed 4 of 5 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[federation/pkg/dnsprovider/providers/coredns/coredns_test.go, line 265 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44403#-Kjq7mPXQ0bsXA3c8lJl:-Kjq7mPXQ0bsXA3c8lJm:b-hxwq2f) ([raw file](https://github.com/kubernetes/kubernetes/blob/73a01968c8b86e313566049660b48c057cb2acfd/federation/pkg/dnsprovider/providers/coredns/coredns_test.go#L265)):*
> ```Go
> /* TestResourceRecordSetsWithMixedTypes  verifies that we can add records with different types.
>    Note: Can not run CommonTestResourceRecordSetsDifferentTypes for coredns, because ResoureRecordSets.List()
>    is not implemented, and Get() API will aggregate rrdata (IPs) for rrsets with the same name, causing
> ```

This comment may not be completely true after this PR #44626. In that PR @madhusudancs has changed Get() to return multiple rrsets for a name with different types. PTAL and update.

---

*[federation/pkg/dnsprovider/tests/commontests.go, line 92 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44403#-Kjq8qQGmFJ0qKds6NbE:-Kjq8qQGmFJ0qKds6NbF:b-n9j8ap) ([raw file](https://github.com/kubernetes/kubernetes/blob/73a01968c8b86e313566049660b48c057cb2acfd/federation/pkg/dnsprovider/tests/commontests.go#L92)):*
> CommonTestResourceRecordSetsWithMixedTypes
The name is ambiguous, gives a feeling to me that the test is for checking multiple record types can be written for the same name. But actually the test is just writing different record types with different names. is this test required at all? 

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44403)*
<!-- Sent from Reviewable.io -->

pmichali commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[federation/pkg/dnsprovider/providers/coredns/coredns_test.go, line 265 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44403#-Kjq7mPXQ0bsXA3c8lJl:-KjsQo_L7B9hk4Z1GbP0:bn67bds) ([raw file](https://github.com/kubernetes/kubernetes/blob/73a01968c8b86e313566049660b48c057cb2acfd/federation/pkg/dnsprovider/providers/coredns/coredns_test.go#L265)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

This comment may not be completely true after this PR #44626. In that PR @madhusudancs has changed Get() to return multiple rrsets for a name with different types. PTAL and update.
</blockquote></details>

Thanks for the update about the other commit. I'll rebase to pick up the changes and update this commit.

---

*[federation/pkg/dnsprovider/tests/commontests.go, line 92 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44403#-Kjq8qQGmFJ0qKds6NbE:-KjsRNwgbdydkbtTHyal:byubgb3) ([raw file](https://github.com/kubernetes/kubernetes/blob/73a01968c8b86e313566049660b48c057cb2acfd/federation/pkg/dnsprovider/tests/commontests.go#L92)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

> CommonTestResourceRecordSetsWithMixedTypes
The name is ambiguous, gives a feeling to me that the test is for checking multiple record types can be written for the same name. But actually the test is just writing different record types with different names. is this test required at all? 
</blockquote></details>

It's been a few weeks since I wrote this, and the PR #44626 may change things, so I'll revisit it. Originally, I wanted to use the CommonTestResourceRecordSetsDifferentTypes test, which had IPv4 and IPv6 IPs for the same name, but coredns didn't support that, so I created this test, using different names.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44403)*
<!-- Sent from Reviewable.io -->

pmichali commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[federation/pkg/dnsprovider/providers/coredns/rrsets.go, line 78 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44403#-KhYmJzTszZTmA_0tUMQ-r1-78:-KjsU3DmnKsNfDgxCoYT:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/c1fb59c3c39c620abf03025bd49272ab6509fbdd/federation/pkg/dnsprovider/providers/coredns/rrsets.go#L78)):*
<details><summary><i>Previously, pmichali (Paul Michali) wrote…</i></summary><blockquote>

Will take a guess at option #2, adding a comment, so that this can move forward.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44403)*
<!-- Sent from Reviewable.io -->

pmichali commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[federation/pkg/dnsprovider/providers/coredns/coredns_test.go, line 265 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44403#-Kjq7mPXQ0bsXA3c8lJl:-KjscRs51jjDMPR8UJja:bw7ez86) ([raw file](https://github.com/kubernetes/kubernetes/blob/73a01968c8b86e313566049660b48c057cb2acfd/federation/pkg/dnsprovider/providers/coredns/coredns_test.go#L265)):*
<details><summary><i>Previously, pmichali (Paul Michali) wrote…</i></summary><blockquote>

Thanks for the update about the other commit. I'll rebase to pick up the changes and update this commit.
</blockquote></details>

Pulling this test method, as existing CommonTestResourceRecordSetsDifferentTypes will work now that Get() has been changed.

---

*[federation/pkg/dnsprovider/tests/commontests.go, line 92 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44403#-Kjq8qQGmFJ0qKds6NbE:-KjscBP6m1TIvSiUjc1A:b1eomzv) ([raw file](https://github.com/kubernetes/kubernetes/blob/73a01968c8b86e313566049660b48c057cb2acfd/federation/pkg/dnsprovider/tests/commontests.go#L92)):*
<details><summary><i>Previously, pmichali (Paul Michali) wrote…</i></summary><blockquote>

It's been a few weeks since I wrote this, and the PR #44626 may change things, so I'll revisit it. Originally, I wanted to use the CommonTestResourceRecordSetsDifferentTypes test, which had IPv4 and IPv6 IPs for the same name, but coredns didn't support that, so I created this test, using different names.
</blockquote></details>

Yes, this is no longer needed, as we can use CommonTestResourceRecordSetsDifferentTypes now, because Get() now returns a list, instead of single rrset. Updating.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44403)*
<!-- Sent from Reviewable.io -->

pmichali commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[federation/pkg/dnsprovider/providers/coredns/coredns_test.go, line 265 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44403#-Kjq7mPXQ0bsXA3c8lJl:-Kjsca6rExTd-R2G44Jt:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/73a01968c8b86e313566049660b48c057cb2acfd/federation/pkg/dnsprovider/providers/coredns/coredns_test.go#L265)):*
<details><summary><i>Previously, pmichali (Paul Michali) wrote…</i></summary><blockquote>

Pulling this test method, as existing CommonTestResourceRecordSetsDifferentTypes will work now that Get() has been changed.
</blockquote></details>

Done.

---

*[federation/pkg/dnsprovider/tests/commontests.go, line 92 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44403#-Kjq8qQGmFJ0qKds6NbE:-Kjscc_PvtlNc2l6dw75:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/73a01968c8b86e313566049660b48c057cb2acfd/federation/pkg/dnsprovider/tests/commontests.go#L92)):*
<details><summary><i>Previously, pmichali (Paul Michali) wrote…</i></summary><blockquote>

Yes, this is no longer needed, as we can use CommonTestResourceRecordSetsDifferentTypes now, because Get() now returns a list, instead of single rrset. Updating.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44403)*
<!-- Sent from Reviewable.io -->

pmichali commented:
Pushed up a new commit, PTAL

shashidharatd commented:
/lgtm
Thanks @pmichali !

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44403#-:-KjvDMrBl5Xs8ttDQmsi:bxcg5gj)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
@madhusudancs, PTAL and approve this PR. Thanks.

pmichali commented:
@shashidharatd thank you for reviewing!

madhusudancs commented:
Reviewed 2 of 5 files at r1, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[federation/pkg/dnsprovider/providers/coredns/rrsets.go, line 78 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44403#-KhYmJzTszZTmA_0tUMQ-r1-78:-KkDTz39Uf46rUcQUhM5:bgab0zj) ([raw file](https://github.com/kubernetes/kubernetes/blob/c1fb59c3c39c620abf03025bd49272ab6509fbdd/federation/pkg/dnsprovider/providers/coredns/rrsets.go#L78)):*
<details><summary><i>Previously, pmichali (Paul Michali) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I would have preferred 3 for clarity and/or readability, but this is fine for now.

---

*[federation/pkg/dnsprovider/tests/commontests.go, line 34 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44403#-KkDVJGQOnrKmFhgar7_:-KkDVJGRGPB9ETrxKEyP:bvj1a73) ([raw file](https://github.com/kubernetes/kubernetes/blob/48caf95a6cf3e77cba6797525b46e2f9733a0132/federation/pkg/dnsprovider/tests/commontests.go#L34)):*
> ```Go
> 	rrset := rrsets.New("alpha.test.com", []string{"8.8.4.4"}, 40, rrstype.A)
> 	addRrsetOrFail(t, sets, rrset)
> 	defer sets.StartChangeset().Remove(rrset).Apply()
> ```

This logic is clever and definitely fewer lines of code, but IMO, it obscures what is really happening. It makes comprehending this code difficult. Why is whatever is returned by `add` is deferred?

I would really urge to revert back to the previous code. Even though, it is verbose, less clever, it is clear.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44403)*
<!-- Sent from Reviewable.io -->

pmichali commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[federation/pkg/dnsprovider/providers/coredns/rrsets.go, line 78 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44403#-KhYmJzTszZTmA_0tUMQ-r1-78:-KkFwxmwE38xfdv7rjBZ:b-69jslt) ([raw file](https://github.com/kubernetes/kubernetes/blob/c1fb59c3c39c620abf03025bd49272ab6509fbdd/federation/pkg/dnsprovider/providers/coredns/rrsets.go#L78)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

I would have preferred 3 for clarity and/or readability, but this is fine for now.
</blockquote></details>

Since I'm changing, I'll modify this.

---

*[federation/pkg/dnsprovider/tests/commontests.go, line 34 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44403#-KkDVJGQOnrKmFhgar7_:-KkFwtV7dw6DEVLRVm0L:bvgtmn8) ([raw file](https://github.com/kubernetes/kubernetes/blob/48caf95a6cf3e77cba6797525b46e2f9733a0132/federation/pkg/dnsprovider/tests/commontests.go#L34)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

This logic is clever and definitely fewer lines of code, but IMO, it obscures what is really happening. It makes comprehending this code difficult. Why is whatever is returned by `add` is deferred?

I would really urge to revert back to the previous code. Even though, it is verbose, less clever, it is clear.
</blockquote></details>

Sure. Will change this back and post an update.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44403)*
<!-- Sent from Reviewable.io -->

pmichali commented:
PTAL I pushed a new version of the changes to revert the changes to addRrsetOrFail(), change the switch statement, and added UT code to check the Zone() method, which was not covered. Rebased this morning (hopefully that is recent enough).

pmichali commented:
PTAL as I pushed up a new version that modifies the switch statement, reverts the addRrsetOrFail() changes to not return a method to clean up. Left the logging for success, to reduce some duplication, and added a bit of code to cover the Zone() method for coverage.

---

Review status: 2 of 5 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44403#-:-KkH4vwzmdB9kQwCMdZk:b-sh1w1d)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
Thanks for making  these changes!

/lgtm

---

Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44403#-:-KkIGqxA4GoLxbmfBp5m:bd4hw8k)*
<!-- Sent from Reviewable.io -->

pmichali commented:
@madhusudancs @shashidharatd Thanks for the prompt review and approval!
